File: TPII-2025-2/PadraoGoF/Estrutural/Decorator/Decorator01.js
// Componente Notificador:
class ComponenteNotificador{
    enviar(msg){
        console.log(`Enviando Msg: ${msg}`);
    }
}

// Componente Decorador Base:
class DecoradorNotificador{
    constructor(componente){
        this.componente = componente;
    }

    enviar(msg){
        this.componente.enviar(msg);
    }
}

// Componentes Decorador Concreto 1 - SMS:
class NotificadorSMS extends DecoradorNotificador{
    enviar(msg){
        super.enviar(msg);
        console.log(`Enviando Msg por SMS: ${msg}`);
    }
}

// Componentes Decorador Concreto 2 - Whatsapp:
class NotificadorWhatsapp extends DecoradorNotificador{
    enviar(msg){
        super.enviar(msg);
        console.log(`Enviando Msg por Whatsapp: ${msg}`);
    }
}

// Componentes Decorador Concreto 3 - Email:
class NotificadorEmail extends DecoradorNotificador{
    enviar(msg){
        super.enviar(msg);
        console.log(`Enviando Msg por Email: ${msg}`);
    }
}

// USO PELO CLIENTE:
const notificador = new ComponenteNotificador();
const notificadorSMS = new NotificadorSMS(notificador);
const notificadorWhatsapp = new NotificadorWhatsapp(notificadorSMS);

notificadorWhatsapp.enviar("Pedido do Habbi's Saindo... ");

File: TPII-2025-2/PadraoGoF/Estrutural/Decorator/Decorator01.java
// Exemplo de Decorator em Java
// O padrão Decorator permite adicionar funcionalidades a objetos de forma dinâmica, sem alterar sua estrutura original.

// Componente base
interface Notificador {
    void enviar(String msg);
}

// Componente concreto
class ComponenteNotificador implements Notificador {
    public void enviar(String msg) {
        System.out.println("Enviando Msg: " + msg);
    }
}

// Decorador base
abstract class DecoradorNotificador implements Notificador {
    protected Notificador componente;
    public DecoradorNotificador(Notificador componente) {
        this.componente = componente;
    }
    public void enviar(String msg) {
        componente.enviar(msg);
    }
}

// Decorador concreto - SMS
class NotificadorSMS extends DecoradorNotificador {
    public NotificadorSMS(Notificador componente) { super(componente); }
    public void enviar(String msg) {
        super.enviar(msg);
        System.out.println("Enviando Msg por SMS: " + msg);
    }
}

// Decorador concreto - Whatsapp
class NotificadorWhatsapp extends DecoradorNotificador {
    public NotificadorWhatsapp(Notificador componente) { super(componente); }
    public void enviar(String msg) {
        super.enviar(msg);
        System.out.println("Enviando Msg por Whatsapp: " + msg);
    }
}

// Decorador concreto - Email
class NotificadorEmail extends DecoradorNotificador {
    public NotificadorEmail(Notificador componente) { super(componente); }
    public void enviar(String msg) {
        super.enviar(msg);
        System.out.println("Enviando Msg por Email: " + msg);
    }
}

// Uso pelo cliente
class MainDecorator {
    public static void main(String[] args) {
        Notificador notificador = new ComponenteNotificador();
        Notificador notificadorSMS = new NotificadorSMS(notificador);
        Notificador notificadorWhatsapp = new NotificadorWhatsapp(notificadorSMS);
        notificadorWhatsapp.enviar("Pedido do Habbi's Saindo... ");
    }
}
